MVP Scheduler Service (#443)
kubos/kubos: docs/ecosystem/services/scheduler.rst

all of the necessary information for that specific schedule. Different schedules, such as
operational, safe mode, etc, will each have their own schedule files.

Schedules consist of three sections: ``init``, ``one-time``, and ``recurring``. Each section
represents a different type of scheduled task.

Tasks in the ``init`` section will be executed on boot or on schedule change. Each 
task in this section will be specified like so:

.. code-block:: json

    {
        "task-name": {
            "command": "task command"
        }
    }

Tasks in the ``one-time`` section will be executed once at a set time. Each task
in this section will be specified like so:

.. code-block:: json

    {
        "task-name": {
            "time": "time of execution in yyyy-mm-dd-hh:mm:ss format",
            "command": "task command"
        }
    }

Tasks in the ``recurring`` section will be executed on a recurring basis. A ``start-time`` and
``end-time`` are given to supply bounds around the occurance of the task. The task
will occur at the given frequency beginning at the ``start-time``. Each task
in this section will be specified like so:

.. code-block:: json

    {
        "task-name": {

>             "start-time": "start time of execution in yyyy-mm-dd-hh:mm:ss format",

Reviewer: nit: date/time format should follow [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601#Combined_date_and_time_representations), so the delimiter between `dd` and `hh` should be either a `T` or a space

            "end-time": "end time of execution in yyyy-mm-dd-hh:mm:ss format",

>             "frequency": "frequency of execution in HH:mm:ss format",

Author: Would it be better to add an extra digit to the hours field or to add a days field?

>             "frequency": "frequency of execution in HH:mm:ss format",

Reviewer: days field would probably be better. It's more human-readable

>             "frequency": "frequency of execution in HH:mm:ss format",

Reviewer 2: To get format ideas from a very full-featured scheduler, skim the readme for https://github.com/jmettraux/rufus-scheduler

>             "command": "task command"

Author: I looked back through our cards and we decided to only implement app execution for MVP, so that cleans this up :sunglasses:.

        }
    }

An example schedule file:

.. code-block:: json

    {
        "init": {
            "start-app-service": {
                "command": "/etc/init.d/S99-app-service start"
            }
        },
        "one-time": {
            "deploy-solar": {
                "time": "2019-08-11-15:20:10"
                "command": "deploy-solar-panels"
            }
        },
        "recurring": {
            "clean-logs-every-12hrs": {
                "start-time": "2019-08-11-15:20:10"
                "end-time": "2019-08-12-15:20:10"
                "frequency": "12:00:00",
                "command": "rm -rf /tmp/logs/*"
            }
        }

>     }

Author: Yes that would be great.

>     }

Reviewer 3: It should also check to see if the apps you're telling it to start are present if possible

The scheduler has the following available configuration parameter which may be 
specified in the ``config.toml`` file under ``[scheduler-service]``.

- ``schedules-dir`` - (Default: ``/home/system/kubos/schedules/``) The path to the
directory where schedules will be stored. This directory will be created if it does
not already exist.

The scheduler also has the standard GraphQL interface parameters available for
configuration under ``[scheduler-service.addr]``.

- ``ip`` - The IP address of the GraphQL server
- ``port`` - The port the GraphQL server will listen on

GraphQL API
-----------

Queries
~~~~~~~

The scheduler exposes a single query, ``schedule``, which exposes information about
the currently active schedule. The ``schedule`` query has the following schema::

    {
        schedule {
            contents: String,
            path: String,
            name: String,
            timeRegistered: String
        }
    }

Mutations
~~~~~~~~~

The scheduler has two mutations: ``switchSchedule`` and ``registerSchedule``. 

The ``switchSchedule`` mutation allows the scheduler to switch between two different schedule modes.
It has the following schema::

    mutation {

>         switchSchedule(name: String!): {

Author: :+1: for name

            success: Boolean!
        }
    }

> The ``registerSchedule`` mutation allows the scheduler to register a new schedule file. It has

Reviewer: nit: You're not going to be registering anything else, so this mutation could just be `register`

Scheduler
=========

The KubOS system includes a scheduler service to facilitate recurring tasks
and one time tasks with specific timing requirements.

Behavior
--------

Upon boot, or service start, the scheduler will read the active schedule file and
load the schedule into memory. The default active schedule file will be found at

> ``/home/system/etc/schedules/active.json``. Any ``init`` tasks will be performed

Reviewer 3: What order will the init tasks be completed in?

Author: I was thinking the order that they are listed in the schedule file, but also open to other ideas.

Reviewer 3: Order in the schedule file is good, just make sure that's documented.

> immediately. All other ``one-time`` or ``recurring`` tasks will be scheduled 

Reviewer 3: Will recurring start immediately, or after their first period? Can I specify a delay time after schedule entry to start recurring?

Author: Because we are specifying a `startTime` and `frequency` for recurring tasks, I was thinking that we would schedule based on those two items. Recurring tasks won't start immediately, but at the end of the first period which hasn't already occurred.

Reviewer 3: I think for recurring, a relative delay time might actually be more appropriate than a start and end. I don't think start and end will almost ever get used

to run at their designated times.

> By default the scheduler will have two schedules: ``operational`` and ``safemode``. These

Reviewer 3: I'm actually onboard with @CatGarab with the modes vs names. I think having many named schedules is fine. I realized what we need more than "modes" generally, is the ability to specify a "safemode" schedule that is used as a fallback if the active schedule can't be read, and can be stored redundantly.

Author: This makes sense. Should we allow the specification of a safemode directory of schedules, or just a singular schedule file?

Reviewer 3: Are we supporting a directory currently?

schedules are represented by different schedule files maintained in the schedules directory.
Only one schedule can be active at any given time.

> The scheduler will have its own log file, ``/var/log/kubos-schedule.log``, which

Reviewer 3: Now that I think about it - should all core services have their own logs?

Reviewer 2: Does this use syslog?

Author: Currently we specify that all services and apps use `syslog` to collect logging. Individual logs for each service would make them more human readable without the need for grepping and parsing. It would also make it quicker to downlink logs specific to one part of the system. 

However if we instrument our logs with the correct information, then it would be simple to parse out the specific service logs from a singular collective log file. The one collective log file  also makes it a bit easier to see the whole state of the system around a certain time, which could be useful for diagnostics. On the other hand, if the logs are meant to be digested, reconstructed, and viewed through a tool like MT, then it does not really matter how we package them onboard.

Reviewer 3: @plauche I think you're right that we should just dump the scheduler logs in with the rest of the services. If we need them parsed out, we can do so after the fact.

will log all schedule related actions the scheduler takes.

Schedule Specification
----------------------

Schedules will be specified in files in the `json` format. Each schedule file will contain
all of the necessary information for that specific schedule. Different schedules, such as
operational, safe mode, etc, will each have their own schedule files.

> Schedules consist of three sections: ``init``, ``one-time``, and ``recurring``. Each section

Reviewer 2: Idea: should there be any `unload` items to run when leaving a schedule?

Author: Hmmm good question. @jacoffey3 is this a use case people expect?

Reviewer 3: So I'd thought about this, and I don't think it's necessary for MVP. I agree that it's something that people might want, but I think the use case can also be satisfied by tackling those items when entering a new schedule. 

Generally, I assume people will kill all running apps when entering safemode to start from a zero state. The only thing that's not tackled with that is if they run another long running executable on init. But, they should put logic in to kill that for their safemode. I'd rather depend on my safemode schedule forcing the system into the appropriate state than trusting others to do proper unloading.

represents a different type of scheduled task.

> Tasks in the ``init`` section will be executed on boot or on schedule change. Each 

Reviewer 3: Maybe here is a better place to describe how init tasks will be executed. Adding a init delay will be necessary I imagine, but there is a possibility that it's too complicated for MVP

Reviewer 2: I think init delay will be needed to allow for relative time command sequences.

Author: Hmm maybe we should consider sequence related functionality in a post-MVP PR and have a discussion about different ways to accomplish that.

task in this section will be specified like so:

.. code-block:: json

    {
        "task-name": {

>             "app-name": "task app",

Reviewer 3: preceeding everything with "app" is probably not necessary. Maybe make this: 
```
{
  "app": {
    "name": "task app",
    "args": "args",
    "config": "path to config"
  }
  schedule fields...
}
```

Author: Yup I agree. Originally I was writing with app and non-app execution in mind, but if we are always running apps then prepending with `app` is not needed.

            "app-args": "optional app args",
            "app-config": "optional path to app config"
        }
    }

Tasks in the ``one-time`` section will be executed once at a set time. Each task
in this section will be specified like so:

.. code-block:: json

    {
        "task-name": {

>             "time": "time of execution in yyyy-mm-dd hh:mm:ss format",

Reviewer 3: For instance, like I mentioned before: 
```
{
  "app": {
    "name": "task app",
    "args": "args",
    "config": "path to config"
  },
  "time": "yyyy-mm-dd hh:mm:ss"
}
```

This way, we can add cmd line commands later with: 

```
{
  "cmd": "executable to run with args",
  "time": "yyyy-mm-dd hh:mm:ss"
}
```

and we will just have to detect if `app` or `cmd` are present.

Author: This seems pretty reasonable.

            "app-name": "task app",
            "app-args": "optional app args",
            "app-config": "optional path to app config"
        }
    }

> Tasks in the ``recurring`` section will be executed on a recurring basis. A ``start-time`` and

Reviewer 2: I'd suggest that start and end should be optional, in which case it just goes forever. Also makes the system work even if the clock is wrong.

Author: That is fine. If the `start-time` is not included, then the recurring intervals will begin when the scheduler is started?

Reviewer 3: This is another point where having a delay would be best. After X time start recurring task.

``end-time`` are given to supply bounds around the occurance of the task. The task
will occur at the given frequency beginning at the ``start-time``. Each task
in this section will be specified like so:

.. code-block:: json

    {
        "task-name": {
            "start-time": "start time of execution in yyyy-mm-dd hh:mm:ss format",
            "end-time": "end time of execution in yyyy-mm-dd hh:mm:ss format",
            "frequency": "frequency of execution in HH:mm:ss format",
            "app-name": "task app",
            "app-args": "optional app args",
            "app-config": "optional path to app config"
        }
    }

An example schedule file:

.. code-block:: json

    {
        "init": {
            "start-camera": {
                "app-name": "activate-camera"
            }
        },
        "one-time": {
            "deploy-solar": {

>                 "time": "2019-08-11 15:20:10"

Reviewer 2: Maybe this is a silly question, but what happens if the clock on the satellite is wrong?

Reviewer 3: Then it doesn't run it. Our core assumption is that system time is our source of truth. We are proceeding as if it's right. We need to account for if it's wrong in other ways.

Scheduler
=========

The KubOS system includes a scheduler service to facilitate recurring

> and one time tasks with specific timing requirements.

Reviewer 3: ```suggestion
and one time tasks with specific timing requirements, 
and to facilitate managing these tasks in groups of system-level modes. 
```

> Behavior

Reviewer 3: This is a really huge wall of text, I'm suggesting some subsections for it rather than just paragraphs.

--------

> The behavior of the scheduler is defined by three levels of organization: *mode*,

Reviewer 3: ```suggestion
Terminology
___________

The behavior of the scheduler is defined by three levels of organization: *mode*,
```

*schedule*, and *task*. The *mode* determines which schedule directory is being executed.
The *schedule* is then assembled by combining all the *task lists* in the schedule

> directory, each of which can contain one or many *tasks*. New *tasks* can be added to

Reviewer 3: Make this a new paragraph to break up the wall of text. 
```suggestion
directory, each of which can contain one or many *tasks*. 

Adding Tasks
____________

New *tasks* can be added to
```

a *schedule* by uploading a *task list* to a location which is accessible to the 
scheduler, and importing that *task list* to the appropriate mode. *Tasks* can
also be added directly via the ``importRawTaskList`` GraphQL mutation.

> Upon boot, or service start, the scheduler reads all task lists in the active 

Reviewer 3: ```suggestion
Bootup Behavior
_______________

Upon boot, or service start, the scheduler reads all task lists in the active 
```

mode directory and schedules all tasks from that mode. The default active mode directory
is found at ``/home/system/etc/schedules/active``, which is a symlink
to the actual active mode's directory.

> Schedules consist of many tasks, each of which has a scheduled execution time.

Reviewer 3: This is a redundant section. It should be in the Task description in the next section

Tasks are loaded into the scheduler when the scheduler service starts,
when a mode becomes active, or when a new task list is imported into the active mode.
Any tasks configured with a ``delay`` will be scheduled for execution
when their corresponding delay has passed. All other tasks configured with a
``time`` or ``period`` will be scheduled to run at their designated times.

> By default the scheduler comes with a single mode: ``safe``. This mode is reserved as a

Reviewer 3: ```suggestion
Scheduler Modes
_______________

By default the scheduler comes with a single mode: ``safe``. This mode is reserved as a
```

fallback and default operating mode. Additional modes can be created and made active
through the GraphQL interface. All modes are represented as different directories in
the schedules directory (``/home/system/etc/schedules``).
Only one mode can be active at any given time.

> The scheduler has its own log file, ``/var/log/kubos-schedule.log``, which

Reviewer 3: This is not true any more, correct?

Author: Correct!

logs all schedule related actions the scheduler takes.

> 

Reviewer 3: Add a new section describing when we go into ``safe`` mode automatically, and title it accordingly

.. _schedule-specification:

> Schedule Specification

Reviewer 3: Maybe this section would be better titled "How to build a Task List" 

It honestly looks a lot like things that would fall under "Behavior" as well, as you're describing how `delay`, `period`, and `time` work, and what Tasks are able to do.

Author: Yes, now that I look over it again there is a fair amount of format specification and behaviour mixed together. Maybe we need separate `Service Behaviour` and `Task Behaviour` sections for the higher level service operations and the more fine grained scheduling behaviour?

Reviewer 3: Yeah that makes sense! Maybe just a "Tasks and How to Make Them" Section? Tells you what a task is, how it behaves, and how to build a Task List

----------------------

Schedules are made up of tasks, which are specified through task lists in the 
`json` format. Each task list contains all of the necessary information for each
task to be scheduled. Multiple task lists can coexist in the same mode folder,
allowing for easy loading of new scheduled tasks.

Tasks can have their scheduled time of execution specified using three different

> fields: ``delay``, ``time``, and ``period``. The ``delay`` field specifies

Reviewer 3: Are any of these mutually exclusive?

            "config": "Optional path to app config",
        }
    }

Tasks configured with a ``time`` field will be executed once at a set time. Each 
one time task is specified like so:

.. code-block:: json

    {
        "name": "Descriptive task-name",
        "time": "Required time of execution in yyyy-mm-dd hh:mm:ss format",
        "app": {
            "name": "Required registered name of app to run",
            "args": ["Optional", "command", "line", "app", "args"],
            "config": "Optional path to app config"
        }
    }

Tasks configured with a ``period`` field will be executed on a recurring basis. The task
will occur at the given ``period`` beginning after the ``delay`` has expired.
Each recurring task in this section is specified like so:

.. code-block:: json

    {
        "name": "Descriptive task-name",
        "delay": "Required start delay in Xh Ym Zs format",
        "period": "Required period of execution in Xh Ym Zs format",
        "app": {
            "name": "Required registered name of app to run",
            "args": ["Optional", "command", "line", "app", "args"],
            "config": "Optional path to app config"
        }
    }

> An example schedule config:

Reviewer 3: ```suggestion
An example task list:
```

Scheduler
=========

The KubOS system includes a scheduler service to facilitate recurring
and one time tasks with specific timing requirements, and to facilitate
managing these tasks in groups of system-level modes. 

Behavior
--------

> Terminology

Reviewer 3: You should add a diagram to illustrate this.

~~~~~~~~~~~

The behavior of the scheduler is defined by three levels of organization: *mode*,
*schedule*, and *task*. The *mode* determines which schedule directory is being executed.
The *schedule* is then assembled by combining all the *task lists* in the schedule
directory, each of which can contain one or many *tasks*. 

Adding Tasks
~~~~~~~~~~~~

New *tasks* can be added to a *schedule* by uploading a *task list* to a location
which is accessible to the scheduler, and importing that *task list* to the appropriate

> mode. *Tasks* can also be added directly via the ``importRawTaskList`` GraphQL mutation.

Reviewer 3: Overly precise for this section of the doc. I don't want them going to look it up.
```suggestion
mode. *Tasks* can also be added directly via GraphQL interface. 
```

Reviewer 2: via *the* GraphQL interface.

Bootup Behavior
~~~~~~~~~~~~~~~

Upon boot, or service start, the scheduler reads all task lists in the active 
mode directory and schedules all tasks from that mode. The default active mode directory
is found at ``/home/system/etc/schedules/active``, which is a symlink
to the actual active mode's directory.

Scheduler Modes
~~~~~~~~~~~~~~~

By default the scheduler comes with a single mode: ``safe``. This mode is reserved as a
fallback and default operating mode. Additional modes can be created and made active
through the GraphQL interface. All modes are represented as different directories in
the schedules directory (``/home/system/etc/schedules``).
Only one mode can be active at any given time.

Failover Behavior
~~~~~~~~~~~~~~~~~

In addition to using the ``safe`` mode as the default operating mode, the scheduler is 
designed to automatically failover to the ``safe`` mode when any problems are detected.
The following situations will cause the scheduler to activate the ``safe`` mode::

    - The scheduler starts and no modes exist
    - The scheduler starts and no mode is active

>     - The scheduler attempts to activate a mode and the mode does not exist

Reviewer 3: Did we talk about if the scheduler is unable to parse a task list? Or if the task list fails checksum?

Author: We have not taken that into consideration with the fail over coverage. We also don't take into consideration the situation where apps in the current mode fail to start and return errors through the `startApp` mutation.

Reviewer 3: What happens? That behavior should be documented.

                }
            },
            {
                "description": "Deploys solar panels",
                "time": "2019-08-11 15:20:10",
                "app": {
                    "name": "deploy-solar-panels"
                }
            },
            {
                "description": "Regular log cleanup",
                "delay": "1h",
                "period": "12h",
                "app": {
                    "name": "clean-logs"
                }
            }
        ]
    }

Specifying Time of Execution
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Tasks can have their scheduled time of execution specified using three different
fields: ``delay``, ``time``, and ``period``. The ``delay`` field specifies
a delay before the task executes. The ``time`` field specifies a time
when the task will be executed. The ``period`` field indicates the app should
be executed on a recurring basis and specifies the period of recurrence. The ``delay``
field is required, except when using the ``time`` field. The ``time`` and ``period``
fields may not be used together.

Delayed Tasks
~~~~~~~~~~~~~

Tasks configured with only a ``delay`` will be executed on boot or on schedule change.

> Tasks will be assigned to the scheduler in an unpredictable order. The actual execution time

Reviewer 3: You restate this in the next line. 
```suggestion
The actual execution time
```

modes. It has the following schema::

    {
        availableModes(name: String): [
            {
               name: String,
               path: String,
               lastRevised: String,
               active: Boolean
               schedule: [TaskList],
            }
        ]
    }

Schemas for Task and Lists
~~~~~~~~~~~~~~~~~~~~~~~~~~

The ``TaskList`` object exposes metadata about individual task lists. It
has the following schema::

    {
        TaskList:
        {
            filename: String,
            path: String,
            timeImported: String
            tasks: [Task],
        }
    }

The ``Task`` object, and it's sub-objects, expose information about
individual schedule tasks. They have the following schemas::

    {
        Task:
        {
            description: String,
            delay: String,
            time: String,
            period: String,
            app: App
        }

        App:
        {
            name: String,
            args: [String],
            config: String,
        }
    }

Mutations
~~~~~~~~~

The scheduler also exposes the following mutations: ``createMode``, ``removeMode``,

> ``activateMode``, ``importTaskList``, and ``removeTaskList``.

Reviewer 3: ```suggestion
``activateMode``, ``importTaskList``, ``removeTaskList``, and ``safeMode``.
```

Creating Modes
~~~~~~~~~~~~~~

The ``createMode`` mutation instructs the scheduler to create a new empty schedule mode.
It has the following schema::

    mutation {
        createMode(name: String!) {
            success: Boolean,
            errors: String
        }
    }

Removing Modes
~~~~~~~~~~~~~~

The ``removeMode`` mutation instructs the scheduler to delete an existing mode's
directory and all schedules within. It cannot be applied to the currently active
mode, or to the *safe* mode. It has the following schema::

    mutation {
        removeMode(name: String!) {
            success: Boolean,
            errors: String
        }
    }

Activating Modes
~~~~~~~~~~~~~~~~

The ``activateMode`` mutation instructs the scheduler to make the specified mode
active. It cannot be used to activate safe mode, the ``safeMode`` mutation is the

> only way to activate safe mode. It has the following schema::

Reviewer 3: Indefinite "it"
```suggestion
only way to activate safe mode.``activeMode`` has the following schema::
```

    mutation {
        activateMode(name: String!): {
            success: Boolean,
            errors: String
        }
    }

Activating Safe Mode
~~~~~~~~~~~~~~~~~~~~

The ``safeMode`` mutation instructs the scheduler to make the *safe* mode
active. This mutation is the only way to activate *safe* mode and can only
activate that mode. It has the following schema::

    mutation {

>         safeMode(name: String!): {

Reviewer 3: Does it need the `name` arg? Can we remove it? 
```suggestion
        safeMode(): {
```

Author: Ahh that's a copy/paste error, that arg doesn't actually exist.

            success: Boolean,
            errors: String
        }
    }

Importing Task Lists
~~~~~~~~~~~~~~~~~~~~

The ``importTaskList`` mutation allows the scheduler to import a new task list into
a specified mode. If the targeted mode is active, all tasks in the task list will be

> immediately loaded for scheduling. It has the following schema::

Reviewer 3: ```suggestion
immediately scheduled. It has the following schema::
```

    mutation {
        importTaskList(path: String!, name: String!, mode:String!): {
            success: Boolean,
            errors: String
        }
    }

Removing Task Lists
~~~~~~~~~~~~~~~~~~~

The ``removeTaskList`` mutation allows the scheduler to remove a task list from
a specified mode. If the mode is active, all tasks in the task list will be removed

> from the scheduler. It as the following schema::

Reviewer 3: ```suggestion
from the schedule. It as the following schema::
```

when the task will be executed. The ``period`` field indicates the app should
be executed on a recurring basis and specifies the period of recurrence. The ``delay``
field is required, except when using the ``time`` field. The ``time`` and ``period``
fields may not be used together.

Delayed Tasks
~~~~~~~~~~~~~

Tasks configured with only a ``delay`` will be executed on boot or on schedule change.
The actual execution time of the task will be affected by the associated delay times.
If more than one task has the exact same delay, the execution order might be unpredictable.
Each ``delay`` task is specified like so:

.. code-block:: json

    {
        "description": "Task description",
        "delay": "Required start delay in Xh Ym Zs format"
        "app": {
            "name": "Required registered name of app to run",
            "args": ["Optional", "command", "line", "app", "args"],
            "config": "Optional path to app config",
        }
    }

One Time Tasks
~~~~~~~~~~~~~~

Tasks configured with a ``time`` field will be executed once at a set time. Each 
one time task is specified like so:

.. code-block:: json

    {
        "description": "Task description",

>         "time": "Required time of execution in yyyy-mm-dd hh:mm:ss format",

Reviewer 2: Will this always be UTC?

Reviewer 2: If so, document expectation of UTC everywhere.

        "app": {
            "name": "Required registered name of app to run",
            "args": ["Optional", "command", "line", "app", "args"],
            "config": "Optional path to app config"
        }
    }

Recurring Tasks
~~~~~~~~~~~~~~~

Tasks configured with a ``period`` field will be executed on a recurring basis. The task

> will occur at the given ``period`` beginning after the ``delay`` has expired.

Author: Ah yes, that reads much better.

mode directory and schedules all tasks from that mode. The default active mode directory
is found at ``/home/system/etc/schedules/active``, which is a symlink
to the actual active mode's directory.

Scheduler Modes
~~~~~~~~~~~~~~~

By default the scheduler comes with a single mode: ``safe``. This mode is reserved as a
fallback and default operating mode. Additional modes can be created and made active
through the GraphQL interface. All modes are represented as different directories in
the schedules directory (``/home/system/etc/schedules``).
Only one mode can be active at any given time.

Failover Behavior
~~~~~~~~~~~~~~~~~

In addition to using the ``safe`` mode as the default operating mode, the scheduler is 
designed to automatically failover to the ``safe`` mode when any problems are detected.
The following situations will cause the scheduler to activate the ``safe`` mode:

    - The scheduler starts and no modes exist
    - The scheduler starts and no mode is active
    - The scheduler attempts to activate a mode and the mode does not exist
    - The scheduler loads an active mode and encounters an invalid task list

> The following situations will **not** cause the scheduler to activate ``safe`` mode:

Reviewer 3: I guess it would have been better to say we need to highlight the fact that we aren't attempting to recover automatically from an app not starting properly. That's all that needs highlighting. 

```suggestion
The only error situation that will **not** cause the scheduler to activate ``safe`` mode is if the scheduler encounters errors when querying the application service to start an app. 
```

>     - The scheduler loads an active mode and encounters a task scheduled for the past

Reviewer 3: Not worth including, as it should be obvious. 
```suggestion
```

>     - The scheduler encounters errors when querying the application service to start an app

Reviewer 3: Removing due to the wording change I suggested above. 
```suggestion
```

>     - The scheduler imports a task list into an active mode and is unable to parse or validate the list

Reviewer 3: Not worth mentioning here, as we mention that we validate tasks on import. 
```suggestion
```
